src/types.rs

```diff
@@ -0,0 +1,559 @@
+            self.len()
+        }
+
+        fn __setitem__(&mut self, key: PyObject, value: PyObject) -> PyResult<()> {
+            Python::with_gil(|py| self.set_item(py, key, value))
+        }
+
+        fn __delitem__(&mut self, key: PyObject) -> PyResult<()> {
+            Python::with_gil(|py| self.remove_item(py, key))?;
+            Ok(())
+        }
+
+        fn __getitem__(&self, key: PyObject) -> PyResult<PyObject> {
+            match Python::with_gil(|py| self.get_item(py, key.clone()))? {
+                Some(obj) => Ok(obj),
+                None => Err(PyErr::new::<PyKeyError, _>(key)),
+            }
+        }
+    }
+
+    #[pyclass]
+    struct Items {
+        dict: IntoIter<Hash, BTreeMap<Hash, (PyObject, PyObject)>>,
+        bucket: Option<IntoIter<Hash, (PyObject, PyObject)>>,
+    }
+
+    impl Items {
+        fn new(dict: Dict) -> Self {
+            Self {
+                dict: dict.map.into_iter(),
+                bucket: None,
+            }
+        }
+
+        fn next(&mut self) -> Option<(PyObject, PyObject)> {
+            if let Some(bucket) = &mut self.bucket {
+                if let Some((_, entry)) = bucket.next() {
+                    return Some(entry);
+                }
+            }
+            self.bucket = match self.dict.next() {
+                Some((_, bucket)) => Some(bucket.into_iter()),
+                None => return None,
+            };
+
+            self.next()
+        }
+    }
+
+    #[pyclass]
+    struct Keys {
+        items: Items,
+    }
+
+    impl Keys {
+        fn new(dict: Dict) -> Self {
+            let items = Items::new(dict);
+            Self { items }
+        }
+    }
+
+    #[pyclass]
+    struct Values {
+        items: Items,
+    }
+
+    impl Values {
+        fn new(dict: Dict) -> Self {
+            let items = Items::new(dict);
+            Self { items }
+        }
+    }
+
+    #[pyproto]
+    impl PyIterProtocol for Items {
+        fn __next__(mut slf: PyRefMut<Self>) -> IterNextOutput<(PyObject, PyObject), ()> {
+            match slf.next() {
+                Some(entry) => IterNextOutput::Yield(entry),
+                None => IterNextOutput::Return(()),
+            }
+        }
+    }
+
+    #[pyproto]
+    impl PyIterProtocol for Keys {
+        fn __next__(mut slf: PyRefMut<Self>) -> IterNextOutput<PyObject, ()> {
+            match slf.items.next() {
+                Some((entry, _)) => IterNextOutput::Yield(entry),
+                None => IterNextOutput::Return(()),
+            }
+        }
+    }
+
+    #[pyproto]
+    impl PyIterProtocol for Values {
+        fn __next__(mut slf: PyRefMut<Self>) -> IterNextOutput<PyObject, ()> {
+            match slf.items.next() {
+                Some((_, entry)) => IterNextOutput::Yield(entry),
+                None => IterNextOutput::Return(()),
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    pub struct HashableDict<'py> {
+        py: Python<'py>,
+        dict: Dict,
+    }
```

Comment:
The point is that current implementation of `pythonize` requires structs to contain python gil inside. But you can't have class with non static lifetime. So that is why we need 2 structs here.

Reply:
Can you add this as a comment there?